### PR TITLE
Change only modified files for docker caching

### PR DESCRIPTION
**Pull request by vitalyzhakov**
Hello!
We use CI process, which includes a  `composer install` command.
Every `composer install` execution modifies the dates for the vendor folder and some files.

It blocks docker layer caching system and increase CI JOB time.

This PR allows composer install and docker cache system to run along.

### CODE REVIEW

File: src/Composer/Installer.php [hunk 1]
```diff
@@ -311,13 +311,6 @@ public function run()
             foreach ($localRepo->getPackages() as $package) {
                 $this->installationManager->ensureBinariesPresence($package);
             }
-
-            $vendorDir = $this->config->get('vendor-dir');
-            if (is_dir($vendorDir)) {
-                // suppress errors as this fails sometimes on OSX for no apparent reason
-                // see https://github.com/composer/composer/issues/4070#issuecomment-129792748
-                @touch($vendorDir);
-            }
```
Comment (Seldaek) [inline src/Composer/Installer.php]: This was done to fix https://github.com/composer/composer/issues/2764 - so removing it will break that use case.

Comment (vitalyzhakov) [reply src/Composer/Installer.php]: If I correctly understand #2764, @mtibben changed composer.lock file via `composer update` command.
Ok, I fix this situation using `touch` only if `composer update` operation.

Comment (Seldaek) [reply src/Composer/Installer.php]: I don't know how you understood this, the word `update` isn't anywhere on that issue.. The point was that after someone updated, if you fetch changes, you run install, and the vendor dir gets modified. What we could do is only update the timestamp if anything changed in the directory / if any operations were executed.

Comment (vitalyzhakov) [reply src/Composer/Installer.php]: bq. after someone updated, if you fetch changes, you run install
After some requirements updated, we must use `composer update` command, is it right?

Comment (Seldaek) [reply src/Composer/Installer.php]: If you changed the requirements you have to run update, but if you fetch an updated lock file from somewhere which updated what was installed, you have to run install to sync up your vendor dir with that state.

Comment (vitalyzhakov) [reply src/Composer/Installer.php]: Thank your for fast answer!
How about use `touch` only after operations?

`Nothing to install or update` message means, that nothing changed, we can inverse it.

Comment (Seldaek) [reply src/Composer/Installer.php]: Yup that looks better.

**Comment by vitalyzhakov** [comment]
@Seldaek , hello!
Do you have any comments on this PR?

File: src/Composer/Autoload/AutoloadGenerator.php [hunk 2]
```diff
@@ -1005,11 +1014,42 @@ protected function sortPackageMap(array $packageMap)
      */
     protected function safeCopy($source, $target)
     {
-        $source = fopen($source, 'r');
-        $target = fopen($target, 'w+');
+        if (!file_exists($target) OR !file_exists($source) OR !$this->filesAreEqual($source, $target)) {
```
Comment (Seldaek) [inline src/Composer/Autoload/AutoloadGenerator.php]: This should be || not OR.

Comment (vitalyzhakov) [reply src/Composer/Autoload/AutoloadGenerator.php]: Ok, fixed

**Comment by vitalyzhakov** [comment]
@Seldaek , hello!
Do you have any comments on this PR?

Or can your merge it to master branch?

**Comment by Seldaek** [comment]
I merge when I have time thank you please refrain from asking every day it just wastes more of my time.

**Comment by Seldaek** [comment]
Thanks!

**Pull request merged**